python/paddle/fluid/tests/unittests/test_backward.py

```diff
@@ -16,72 +16,173 @@
-    label = fluid.layers.cast(label, dtype='int64')
-    out = fluid.layers.one_hot(input=label, depth=100)
-    loss = fluid.layers.mean(out)
-    return loss
-
-
-def case4_with_no_grad_op_maker():
-    out = fluid.layers.gaussian_random(shape=[20, 30])
-    loss = fluid.layers.mean(out)
-    return loss
+import numpy as np
+
+
+class BackwardNet(object):
+    def __init__(self):
+        self.stop_gradient_grad_vars = []
+        self.no_grad_vars = []
+        self.params_names = []
+        self.op_path = []
+
+    def build_model(self):
+        raise NotImplementedError
+
+    def init_data(self):
+        raise NotImplementedError
+
+    def check_backward(self, loss, main_program):
+        params_grads = self.check_params_grad(loss)
+        # 1.1 get_stop_gradients
+        no_grad_dict = self.check_stop_gradient(main_program)
+        # 1.2 find_op_path
+        op_path, block_no_grad_set = self.check_op_path(
+            main_program.block(0), [loss], [], no_grad_dict)
+        # 1.3 _find_no_grad_vars
+        no_grad_vars = self.check_find_no_grad_vars(
+            main_program.block(0), op_path, [loss], block_no_grad_set)
+        # update no_grad_dict
+        block_no_grad_set.update(no_grad_vars)
+        no_grad_dict[0].update(
+            list(map(fluid.backward._append_grad_suffix_, block_no_grad_set)))
+
+    def check_params_grad(self, loss, parameter_list=None, no_grad_set=None):
+        params_grads = fluid.backward.append_backward(loss, parameter_list,
+                                                      no_grad_set)
+        params_names = set([var[0].name for var in params_grads])
+        assert len(params_names - set(self.params_names)) == 0
+
+        return params_grads
+
+    def check_stop_gradient(self, program):
+        no_grad_dict = fluid.backward._get_stop_gradients_(program)
+        if no_grad_dict is not None and isinstance(no_grad_dict, dict):
+            assert len(no_grad_dict[0] - set(self.stop_gradient_grad_vars)) == 0
+
+        return no_grad_dict
+
+    def check_op_path(self, root_block, outputs, inputs=[], no_grad_dict=None):
+        if no_grad_dict is None or not isinstance(no_grad_dict, dict):
+            block_no_grad_set = None
+        else:
+            block_no_grad_set = set(
+                map(fluid.backward._strip_grad_suffix_, no_grad_dict[0]))
+        op_path = fluid.backward._find_op_path_(root_block, outputs, inputs,
+                                                block_no_grad_set)
+        op_types = [op.type for op in op_path]
+        assert op_types == self.op_path
+
+        return op_path, block_no_grad_set
+
+    def check_find_no_grad_vars(self, root_block, op_path, targets,
+                                block_no_grad_set):
+        no_grad_vars = fluid.backward._find_no_grad_vars(
+            root_block, op_path, targets, block_no_grad_set)
+        assert len(no_grad_vars) == len(self.no_grad_vars)
```

Comment:
Would you mind also assert the no_grad_vars are same (not only the length)?

Reply:
Fixed.